app/models/service_provider.rb

```diff
@@ -158,15 +158,30 @@ def certs_are_pems
   end
 
   def validate_attribute_bundle
-    if (attribute_bundle || []).empty?
+    # attribute bundle should not be empty when saml and ial2 are selected
+    if ((attribute_bundle || []).empty? && ial == 2 && identity_protocol == 'saml')
       errors.add(:attribute_bundle, 'Attribute bundle cannot be empty')
       return false
     end
 
-    possible_attributes = ALLOWED_IAL1_ATTRIBUTES + ALLOWED_IAL2_ATTRIBUTES
-    return if ial == 2 && (attribute_bundle - possible_attributes).empty?
-    return if (attribute_bundle - ALLOWED_IAL1_ATTRIBUTES).empty?
+    return true if !attribute_bundle
+
+    if contains_invalid_attribute?
+      errors.add(:attribute_bundle, 'Contains invalid attributes')
+      return false
+    end
 
-    errors.add(:attribute_bundle, 'Contains invalid IAL attributes')
+    if ial == 1 && ((attribute_bundle & ALLOWED_IAL2_ATTRIBUTES).length > 0)
```

Comment:
I would change this to:
```suggestion
    if ial == 1 && (attribute_bundle & ALLOWED_IAL2_ATTRIBUTES).present?
```
because it looks like we had `attribute_bundle || []` above which implies `attribute_bundle` might be nil, and:

```
nil & []
=> false
```
and `false` won't have a `#length` method

Reply:
I had a nil check for attribute_bundle on line 167 but this is still good advice and allow me to cleanup that line, too.